clients/src/main/java/org/apache/kafka/clients/admin/internals/RemoveMembersFromConsumerGroupHandler.java

```diff
@@ -79,55 +90,82 @@ public String apiName() {
         Set<CoordinatorKey> groupIds,
         AbstractResponse abstractResponse
     ) {
+        validateKeys(groupIds);
+
         final LeaveGroupResponse response = (LeaveGroupResponse) abstractResponse;
-        Map<CoordinatorKey, Map<MemberIdentity, Errors>> completed = new HashMap<>();
-        Map<CoordinatorKey, Throwable> failed = new HashMap<>();
-        List<CoordinatorKey> unmapped = new ArrayList<>();
+        final Map<CoordinatorKey, Map<MemberIdentity, Errors>> completed = new HashMap<>();
+        final Map<CoordinatorKey, Throwable> failed = new HashMap<>();
+        final Set<CoordinatorKey> groupsToUnmap = new HashSet<>();
+        final Set<CoordinatorKey> groupsToRetry = new HashSet<>();
 
-        final Errors error = Errors.forCode(response.data().errorCode());
+        final Errors error = response.topLevelError();
         if (error != Errors.NONE) {
-            handleError(groupId, error, failed, unmapped);
+            handleGroupError(groupId, error, failed, groupsToUnmap, groupsToRetry);
         } else {
             final Map<MemberIdentity, Errors> memberErrors = new HashMap<>();
             for (MemberResponse memberResponse : response.memberResponses()) {
+                Errors memberError = Errors.forCode(memberResponse.errorCode());
+                String memberId = memberResponse.memberId();
+
                 memberErrors.put(new MemberIdentity()
-                                     .setMemberId(memberResponse.memberId())
+                                     .setMemberId(memberId)
                                      .setGroupInstanceId(memberResponse.groupInstanceId()),
-                                 Errors.forCode(memberResponse.errorCode()));
+                    memberError);
 
             }
             completed.put(groupId, memberErrors);
         }
-        return new ApiResult<>(completed, failed, unmapped);
+
+        if (groupsToUnmap.isEmpty() && groupsToRetry.isEmpty()) {
```

Comment:
This does not seem necessary as we always expect the top level error. Would it make sense to handle it like we did here: https://github.com/apache/kafka/pull/11019/files#diff-e7eafbafe0b75099d0c8b4083c03c653d57245ef7b0fcfae7b9ccd258a9024e3R117?

Reply:
Agree! Updated!